### PR TITLE
[Week4] - jungmin

### PR DESCRIPTION
**Pull request by lego37yoon**
## 활성화 함수(활성 함수)

> activation function

입력 신호의 총합을 출력 신호로 변환하는 함수. 이전 계층에 있는 노드, 인공 뉴런에서 다음 뉴런으로 출력하는 과정에서 사용한다.

주로 선형 변수와 함께 쓰이는 비선형 함수로, 시그모이드, 하이퍼볼릭 탄젠트, ReLU, 리키 ReLU, 맥스아웃, ELU 등의 함수가 있다.

### 시그모이드 함수

대표적인 활성 함수로, 입력 신호의 총합을 0~1 사이의 값으로 바꿔준다. 값이 커질수록 뉴런의 활성화률(firing rate)은 1, 작아질수록 0으로 수렴한다.

### ReLU

> Rectified Linear Unit

연속 함수를 사용했기 때문에 그래디언트 포화 문제에서 벗어나지 못했다는 착안점에서 고안. 입력이 0 이상이면 값을 그대로, 0보다 작으면 0을 출력한다. 따라서 0 이상 구간에서는 수렴하지 않으며, 단순 출력이므로 계산 속도가 빠르다.

## 배치 정규화

미니 배치 단위로 데이터의 분포가 평균 0, 분산 1이 되도록 정규화한다. 단, 입력 데이터에 대해 정규화하면 대부분 0에 가깝게 되므로 sigmoid 함수에서 선형(linearity) 구간에 빠진다. 이 문제를 위해 scaling, shifting을 적용(이 값에 대한 가중치, 편향)한다.

## 신경망 설계의 주요 요소

은닉층이 하나라도 충분한 뉴런 수를 가지면 이론상 복잡한 함수도 근사 가능하지만, 심층 신경망은 파라미터 효율성이 높아 더 적은 뉴런으로도 복잡한 문제 해결이 가능하다.

한편, 뉴런 수는 첫 번째 은닉 층에 많은 뉴런을 두고 점차 줄이는 구조가 효과적인 것으로 알려져 있으며 과대 적합을 방지하기 위해 규제 기법을 사용한다.

### 학습률 (Learning Rate)

가장 중요한 하이퍼 파라미터로 학습률이 너무 크면 발산, 너무 작으면 학습 속도가 느려지는 경향이 있다.

그래서 최적의 학습률은 손실 그래프에서 손실이 급격히 감소하기 시작하는 지점의 약 10배 낮은 값이다.

### 배치 크기

작으면 더 좋은 일반화 성능을, 크면 훈련 속도 향상을 보인다.

### CODE REVIEW

**Review by kimhamyong — APPROVED** [review]
task1: 잘 수행하셨습니다.

**Comment by kimhamyong** [comment]
task1:

현재 케라스에선 `Dense(256, input_shape=(784,))` 대신 `Input` 레이어를 사용하는 것이 더 권장되고 있습니다.
   ```python
   from tensorflow.keras.layers import Input
   model = Sequential([
       Input(shape=(784,)),
       Dense(256),
       ...
   ])
   ```

**Comment by kimhamyong** [comment]
task1:

훈련 데이터의 일부를 검증 데이터로 사용해 훈련 및 검증 정확도를 비교할 수 있습니다.

   ```python
model.fit(x_train, y_train, batch_size=64, epochs=5, validation_split=0.1)
   ```

**Comment by kimhamyong** [comment]
task1: 

추가적으로, 학습 도중 검증 손실이 더 이상 감소하지 않을 때 학습을 조기 종료하도록 EarlyStopping 콜백을 추가하면 불필요한 에포크 실행을 방지할 수 있습니다. (필수 사항 X)

   ```python
from tensorflow.keras.callbacks import EarlyStopping
early_stopping = EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True)
history = model.fit(x_train, y_train, batch_size=64, epochs=50, validation_split=0.2, callbacks=[early_stopping])
   ```

**Review by giyomdabin — APPROVED** [review]
task2: 주어진 과제에 대해 잘 수행하셨습니다. 수고하셨습니다!
